pkg/build/controller/build/build_controller.go

```diff
@@ -0,0 +1,850 @@
+	}
+	if !exists {
+		glog.V(2).Infof("Build %q has been deleted", key)
+		return nil, nil
+	}
+
+	return obj.(*buildapi.Build), nil
+}
+
+// podUpdated gets called by the pod informer event handler whenever a pod
+// is updated or there is a relist of pods
+func (bc *BuildController) podUpdated(old, cur interface{}) {
+	// A periodic relist will send update events for all known pods.
+	curPod := cur.(*kapi.Pod)
+	oldPod := old.(*kapi.Pod)
+	// The old and new ResourceVersion will be the same in a relist of pods.
+	// Here we ignore pod relists because we already listen to build relists.
+	if curPod.ResourceVersion == oldPod.ResourceVersion {
+		return
+	}
+	if isBuildPod(curPod) {
+		bc.enqueueBuildForPod(curPod)
+	}
+}
+
+// podDeleted gets called by the pod informer event handler whenever a pod
+// is deleted
+func (bc *BuildController) podDeleted(obj interface{}) {
+	pod, ok := obj.(*kapi.Pod)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone: %+v", obj))
+			return
+		}
+		pod, ok = tombstone.Obj.(*kapi.Pod)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a pod: %+v", obj))
+			return
+		}
+	}
+	if isBuildPod(pod) {
+		bc.enqueueBuildForPod(pod)
+	}
+}
+
+// buildAdded is called by the build informer event handler whenever a build
+// is created
+func (bc *BuildController) buildAdded(obj interface{}) {
+	build := obj.(*buildapi.Build)
+	bc.enqueueBuild(build)
+}
+
+// buildUpdated is called by the build informer event handler whenever a build
+// is updated or there is a relist of builds
+func (bc *BuildController) buildUpdated(old, cur interface{}) {
+	build := cur.(*buildapi.Build)
+	bc.enqueueBuild(build)
+}
+
+// enqueueBuild adds the given build to the queue.
+func (bc *BuildController) enqueueBuild(build *buildapi.Build) {
+	key, err := kcontroller.KeyFunc(build)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for build %#v: %v", build, err))
+		return
+	}
+	bc.queue.Add(key)
+}
+
+// enqueueBuildForPod adds the build corresponding to the given pod to the controller
+// queue. If a build is not found for the pod, then an error is logged.
+func (bc *BuildController) enqueueBuildForPod(pod *kapi.Pod) {
+	buildName := buildutil.GetBuildName(pod)
+	build, err := bc.buildStore.Builds(pod.Namespace).Get(buildName)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't retrieve build for pod %s/%s: %v", pod.Namespace, pod.Name, err))
+		return
+	}
+	bc.enqueueBuild(build)
+}
+
+// handleError is called by the main work loop to check the return of the handleBuild or handlePodDeleted
+// calls. If an error occurred, then the key is re-added to the queue unless it has been retried too many
+// times.
+func (bc *BuildController) handleError(err error, key interface{}) {
+	if err == nil {
+		bc.queue.Forget(key)
+		return
+	}
+
+	if strategy.IsFatal(err) {
```

Comment:
I think this isn't right.  FWIW I think we may be underusing the FatalError concept, but more importantly, I think if one bubbles up to here, then it should result in a transition to Error state.